src/Bicep.Core.IntegrationTests/ParamsParserTests.cs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using Bicep.Core.UnitTests.Assertions;
+using Bicep.Core.Parsing;
+using Bicep.Core.Samples;
+using Bicep.Core.Syntax;
+using Bicep.Core.UnitTests.Utils;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System.Linq;
+using System;
+
+namespace Bicep.Core.IntegrationTests
+{
+    [TestClass]
+    public class ParamsParserTests
+    {
+        private class SyntaxCollectorVisitor : SyntaxVisitor
+        {
+            public record SyntaxItem(SyntaxBase Syntax, SyntaxBase? Parent, int Depth);
+
+            private readonly IList<SyntaxItem> syntaxList = new List<SyntaxItem>();
+            private SyntaxBase? parent = null;
+            private int depth = 0;
+
+            private SyntaxCollectorVisitor()
+            {
+            }
+
+            public static ImmutableArray<SyntaxItem> Build(ProgramSyntax syntax)
+            {
+                var visitor = new SyntaxCollectorVisitor();
+                visitor.Visit(syntax);
+
+                return visitor.syntaxList.ToImmutableArray();
+            }
+
+            protected override void VisitInternal(SyntaxBase syntax)
+            {
+                syntaxList.Add(new(Syntax: syntax, Parent: parent, Depth: depth));
+
+                var prevParent = parent;
+                parent = syntax;
+                depth++;
+                base.VisitInternal(syntax);
+                depth--;
+                parent = prevParent;
+            }
+        }
+
+        [NotNull]
+        public TestContext? TestContext { get; set; }
+
+        [DataTestMethod]
+        [DynamicData(nameof(GetData), DynamicDataSourceType.Method, DynamicDataDisplayNameDeclaringType = typeof(DataSet), DynamicDataDisplayName = nameof(DataSet.GetDisplayName))]
+        [TestCategory(BaselineHelper.BaselineTestCategory)]
+        public void Params_Parser_should_produce_expected_syntax(DataSet dataSet)
+        {
+            if (dataSet.BicepParam is null)
+            {
+                return;
+            }
+
+            var program = ParamsParserHelper.ParamsParse(dataSet.BicepParam);
+            var syntaxList = SyntaxCollectorVisitor.Build(program);
+            var syntaxByParent = syntaxList.ToLookup(x => x.Parent);
+
+            string getLoggingString(SyntaxCollectorVisitor.SyntaxItem data)
```

Comment:
We can also make it a public helper method of `SyntaxCollectorVisitor` which both `ParserTests` and `ParamsParserTests` can reuse.

Reply:
+1, would be good to extract common logic here